service/storage_service.cc

```diff
@@ -2588,6 +2587,9 @@ void storage_service::node_ops_cmd_check(gms::inet_address coordinator, const no
             msg = format("node_ops_cmd_check: Node {} rejected node_ops_cmd={} from node={} with ops_uuid={}, pending_node_ops={}, pending node ops is in progress",
                     get_broadcast_address(), req.cmd, coordinator, req.ops_uuid, ops_uuids);
         }
+    } else if (req.cmd == node_ops_cmd::decommission_heartbeat || req.cmd == node_ops_cmd::removenode_heartbeat ||
+            req.cmd == node_ops_cmd::replace_heartbeat || req.cmd == node_ops_cmd::bootstrap_heartbeat) {
+        // We allow node_ops_cmd heartbeat to be sent before prepare cmd
```

Comment:
Why do that at all?
This prevents the coordinator from using the heartbeat cmd to probe whether the the ops_uuid is still valid on other nodes.
It's better to let the coordinator get the error produced here and handle it.
For example:
1. The coordinator could track which node was sent the *_prepare cmd already, and if it hasn't been sent, ignore the "node ops is unknown" error.
2. As described in https://github.com/scylladb/scylladb/issues/12969#issuecomment-1441690914, we can have another step, for safety, the will send a heartbeat update cmd to all participating node before committing the operation.  This hunk renders the *_heartbeat commands useless for this purpose, since the cmd will blindly succeed, without rearming the watchdog timer.

Reply:
With that in mind, I still think we should implement this check on the coordinator side, despite being not strictly necessary for the particular scenario we're dealing with right now.
It is needed as a safety precaution for unforeseen issues that might cause the participating nodes to abort the node operation on their own.

Reply:
We already have a dedicated node_ops_cmd::query_pending_ops to query valid pending ops.  It is best to avoid reuse the heartbeat verb. 

I do not like the idea to special case the handling of heartbeat depending on if a _prepare cmd is sent or not. It adds unnecessary complexity.

Reply:
If the peer node aborts the node operation on their own, the coordinator will know, e.g., the streaming / repair fails, or the following verb fails. We do not really need the heartbeat to check if the peer node is aborting. The heartbeat is used to protect the coordinator goes away.

Reply:
> If the peer node aborts the node operation on their own, the coordinator will know, e.g., the streaming / repair fails, or the following verb fails.

At least with streaming, the coordinator doesn't know that the peer aborted.
That's what we saw in #12969.

> We do not really need the heartbeat to check if the peer node is aborting. The heartbeat is used to protect the coordinator goes away.

I'm ok with using `query_pending_ops` to see if the operation is still alive on the peer.

But even with that, There's no point in hiding the heartbeat update failure on the peer side.
If you think it's not worth warning about, please change the log level on the coordinator side to debug.